Rewrite installation docs to not recommend anaconda / defaults channel (#16555)
astropy/astropy: docs/install.rst

.. _installing-astropy:

************
Installation
************

.. _installing-astropy-with-conda:

Installing Python and Astropy with Conda
****************************************

> There are many ways to install Python, but even if you have Python installed somewhere on your computer we recommend following these instructions anyway.

Reviewer: ```suggestion
There are many ways to install Python, but even if you have Python installed somewhere on your computer we recommend following these instructions.
```

The easiest way to manage a Python installation together with an Astropy installation is with the ``conda`` package manager.
If you already have Python and ``conda`` working you can skip the next section.
However, if you are using the Anaconda Python Distribution, we recommend you still install miniforge as described below.

Installing miniforge
====================

We recommend installing Python with `miniforge <https://github.com/conda-forge/miniforge/#miniforge>`__.
This will install ``conda`` and automatically configure the default channel (a channel is a remote software repository) to be ``conda-forge``.

> First, download the installer for your system and architecture from the links below:

Reviewer 2: I am not sure if we want to maintain a list of installers here. We are not a miniforge documentation. I don't feel like this is our job.

Reviewer 3: also, installers will change and their locations will move...

Reviewer 2: miniforge not having nice doc is not our problem to solve, or if you want to solve that, it isn't here. I am okay with suggesting to use miniforge but I do not want our doc to turn into their doc.

Reviewer 4: > and the miniforge documentation is rubbish to non-existent

@Cadair - Is this what you were talking about?
https://github.com/conda-forge/miniforge/blob/main/README.md

I agree your version is prettier and nicer, but it seems like that README has the basic information on download and install. Maybe it's good enough?

Author: I think it really sucks if we make users click out to an inferior set of documentation (clicking out alone is friction) vs keeping like 6 links in our docs. If we contribute upstream in the future and make things better then sure let's get rid of it.

Author: Also: https://github.com/conda-forge/miniforge/issues/76

Reviewer 4: Since that issue to make a nice download page has been open for nearly 4 years, I think we should just use the nice one that @Cadair has create for astropy. Maybe if this is brought to their attention they will use and maintain it, but in the meantime let's not stall this useful PR!

.. grid:: 3

    .. grid-item-card:: Linux

        `x86-64 <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh>`__

        `aarch64 <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-aarch64.sh>`__

        `ppc64le <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-ppc64le.sh>`__

    .. grid-item-card:: Windows
        :link: https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Windows-x86_64.exe

        `x86-64 <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Windows-x86_64.exe>`__

    .. grid-item-card:: Mac

        `x86-64 <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-x86_64.sh>`__

>         `arm64 (Apple Silicon) <https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh>`__

Reviewer: I'd suggest puttin arm64 first since it's more and more common and should soon become the only used macOS arch (written from my x86-64 2018 macbook)

Then select your platform to install miniforge:

.. tab-set::

    .. tab-item:: Linux & Mac
        :sync: platform

        Linux & Mac Run the script downloaded above, with
        ``bash <filename>``. The following should work:

        .. code-block:: console

            bash Miniforge3-$(uname)-$(uname -m).sh

        Once the installer has completed, close and reopen your terminal.

    .. tab-item:: Windows
        :sync: platform

        Double click the executable file downloaded from
        the links above.

        Once the installer has completed you should have a new "miniforge
        Prompt" entry in your start menu.

In a new terminal (miniforge Prompt on Windows) run ``conda list`` to test that the install has worked.

Using an Existing Conda Installation
====================================

If you already have a conda installation you can use it to install astropy.

> We do not recommend reusing an anaconda installation, as the default packages provided by Anaconda Inc. are often out dated and can require a license for use at a large organisaiton.

Reviewer: ```suggestion
We do not recommend reusing an anaconda installation, as the default packages provided by Anaconda Inc. are often out dated and can require a license for use at a large organisation.
```

> If you have installed miniforge or anaconda you can add ``-c conda-forge`` to all installation commands to install packages from `conda-forge <https://conda-forge.org>`__.

Reviewer: Maybe a good place to also mention how to do it permanently ?
```
conda config --prepend channels conda-forge
```

Author: I got myself stuck in a bit of a brain loop with this. You could write a whole essay on how to unravel yourself out of a defaults / anaconda install and not get shot in the foot.

sunpy has [this](https://docs.sunpy.org/en/stable/tutorial/installation.html#installing-sunpy) here:
```
conda config --add channels conda-forge
conda config --set channel_priority strict
```
which is a good thing to have, but if you have installed miniforge then you don't need to run it, and if you have full-fat anaconda then running them has the potential to really break your base environment. So I am not sure what the best thing to recommend to people who really want to use Anaconda or don't want to install a new miniforge.

Reviewer: Right. We should probably keep it simple and repeatable then. Users who get tired of typing `-c conda-forge` can figure out how to skip it easily enough.

Installing astropy
==================

To install ``astropy``, start by launching a terminal (under a UNIX-like system) or the miniforge Prompt (under Windows).
Now we will create and activate a new virtual environment to install ``astropy`` into:

.. code-block:: bash

>     $ conda create --name astropy

Reviewer 4: Per my comment below about always using `pip` to install astropy:
```suggestion
    $ conda create -c conda-forge --name astropy-env python
```

>     $ conda create --name astropy
>     $ conda activate astropy

Reviewer: To avoid confusing first-time users, I think it might be preferable to use a distinctive name for then environment itself
```suggestion
    $ conda create --name astropy-env
    $ conda activate astropy-env
```

> In this case the environment we have created is named 'astropy'.

Reviewer: ```suggestion
In this case the environment we have created is named 'astropy-env'.
```

Feel free to change this to a different environment name.

The benefit of using a virtual environment is that it allows you to install packages without affecting any other Python installation on your system.
This also means you can work on multiple projects (research or coding) with different package requirements without them interfering with each other.

Now we have a fresh environment we can install ``astropy``:

.. code-block:: bash

>     $ conda install astropy

Reviewer 4: At this point I think we should recommend the `pip install` path for installing astropy. In other words `conda` is great for creating the base Python isolated (or "virtual") environment, but pip works better for installing the rest. In particular it is much easier to specify additional dependencies `recommended` or `all`. `conda install astropy` installs only minimal dependencies and there is no easy way to get the recommended or all dependencies.

Instead, start with:
```
conda create -n astropy-env python
```
Then say "Now install astropy using pip as described in the next section."

Reviewer 3: fwiw, i completely agree with this and have recommended this approach in other contexts for the stated reasons. conda is great for pulling in the proverbial kitchen sink, but if you want to fine-tune what you install, pip is a much better approach.

This will install ``astropy`` and all of its dependencies.
If you want to install another package later, you can run ``conda install <package_name>``.

> You will need to run `conda activate astropy` to activate this environment in new terminals in the future.

Reviewer: ```suggestion
You will need to run `conda activate astropy-env` to activate this environment in new terminals in the future.
```

.. _installing-astropy-with-pip:

Installing ``astropy`` with pip
*******************************

.. warning::

  If you are new to Python and/or do not have familiarity with `Python virtual
  environments <https://docs.python.org/3/tutorial/venv.html>`_, then we recommend

>   starting following the conda instructions above.

Reviewer 4: ```suggestion
  starting with the conda instructions above.
```

If you get a ``PermissionError`` this means that you do not have the required
administrative access to install new packages to your Python installation. In
this case you should create a :ref:`virtual environment <astropy-dev:virtual_envs>`.

.. danger::

>   Do **not** install ``astropy`` or other third-party packages using ``sudo``.

Reviewer: 👍🏻
